lib/Optimizer/Dialect/FIRType.cpp

```diff
@@ -0,0 +1,1294 @@
+    return new (storage) RealTypeStorage{kind};
+  }
+
+  KindTy getFKind() const { return kind; }
+
+protected:
+  KindTy kind;
+
+private:
+  RealTypeStorage() = delete;
+  explicit RealTypeStorage(KindTy kind) : kind{kind} {}
+};
+
+/// Boxed object (a Fortran descriptor)
+struct BoxTypeStorage : public mlir::TypeStorage {
+  using KeyTy = std::tuple<mlir::Type, mlir::AffineMapAttr>;
+
+  static unsigned hashKey(const KeyTy &key) {
+    auto hashVal{llvm::hash_combine(std::get<mlir::Type>(key))};
+    return llvm::hash_combine(
+        hashVal, llvm::hash_combine(std::get<mlir::AffineMapAttr>(key)));
+  }
+
+  bool operator==(const KeyTy &key) const {
+    return std::get<mlir::Type>(key) == getElementType() &&
+           std::get<mlir::AffineMapAttr>(key) == getLayoutMap();
+  }
+
+  static BoxTypeStorage *construct(mlir::TypeStorageAllocator &allocator,
+                                   const KeyTy &key) {
+    auto *storage = allocator.allocate<BoxTypeStorage>();
+    return new (storage) BoxTypeStorage{std::get<mlir::Type>(key),
+                                        std::get<mlir::AffineMapAttr>(key)};
+  }
+
+  mlir::Type getElementType() const { return eleTy; }
+  mlir::AffineMapAttr getLayoutMap() const { return map; }
+
+protected:
+  mlir::Type eleTy;
+  mlir::AffineMapAttr map;
+
+private:
+  BoxTypeStorage() = delete;
+  explicit BoxTypeStorage(mlir::Type eleTy, mlir::AffineMapAttr map)
+      : eleTy{eleTy}, map{map} {}
+};
+
+/// Boxed CHARACTER object type
+struct BoxCharTypeStorage : public mlir::TypeStorage {
+  using KeyTy = KindTy;
+
+  static unsigned hashKey(const KeyTy &key) { return llvm::hash_combine(key); }
+
+  bool operator==(const KeyTy &key) const { return key == getFKind(); }
+
+  static BoxCharTypeStorage *construct(mlir::TypeStorageAllocator &allocator,
+                                       KindTy kind) {
+    auto *storage = allocator.allocate<BoxCharTypeStorage>();
+    return new (storage) BoxCharTypeStorage{kind};
+  }
+
+  KindTy getFKind() const { return kind; }
+
+  // a !fir.boxchar<k> always wraps a !fir.char<k>
+  CharacterType getElementType(mlir::MLIRContext *ctxt) const {
+    return CharacterType::get(ctxt, getFKind());
+  }
+
+protected:
+  KindTy kind;
+
+private:
+  BoxCharTypeStorage() = delete;
+  explicit BoxCharTypeStorage(KindTy kind) : kind{kind} {}
+};
+
+/// Boxed PROCEDURE POINTER object type
+struct BoxProcTypeStorage : public mlir::TypeStorage {
+  using KeyTy = mlir::Type;
+
+  static unsigned hashKey(const KeyTy &key) { return llvm::hash_combine(key); }
+
+  bool operator==(const KeyTy &key) const { return key == getElementType(); }
+
+  static BoxProcTypeStorage *construct(mlir::TypeStorageAllocator &allocator,
+                                       mlir::Type eleTy) {
+    assert(eleTy && "element type is null");
+    auto *storage = allocator.allocate<BoxProcTypeStorage>();
+    return new (storage) BoxProcTypeStorage{eleTy};
```

Comment:
Suggestion: 
`return std::allocator_traits<decltype(allocator)>::construct<BoxProcTypeStorage>(allocator, eleTy);`

Reply:
Thanks for the suggestion, I find the two liner less intimidating.